interfaces/acados_template/acados_template/c_templates_tera/ros2/ocp_node_templates/node.in.cpp

```diff
@@ -0,0 +1,619 @@
+        rclcpp::shutdown();
+    }
+
+    ocp_nlp_config_ = {{ model.name }}_acados_get_nlp_config(ocp_capsule_);
+    ocp_nlp_dims_ = {{ model.name }}_acados_get_nlp_dims(ocp_capsule_);
+    ocp_nlp_in_ = {{ model.name }}_acados_get_nlp_in(ocp_capsule_);
+    ocp_nlp_out_ = {{ model.name }}_acados_get_nlp_out(ocp_capsule_);
+    ocp_nlp_opts_ = {{ model.name }}_acados_get_nlp_opts(ocp_capsule_);
+    RCLCPP_INFO(this->get_logger(), "acados solver created successfully.");
+
+    this->set_cost_weights();
+    this->set_constraints();
+    {%- if has_slack %}
+    void set_slack_weights();
+    {%- endif %}
+
+    RCLCPP_INFO(this->get_logger(), "acados solver initialized successfully.");
+}
+
+void {{ ClassName }}::control_loop() {
+    // TODO: check for received msgs first
+    std::array<double, {{ model.name | upper }}_NX> x0{};
+    {%- if dims.ny_0 > 0 %}
+    std::array<double, {{ model.name | upper }}_NY0> yref0{};
+    {%- endif %}
+    {%- if dims.ny > 0 %}
+    std::array<double, {{ model.name | upper }}_NY> yref{};
+    {%- endif %}
+    {%- if dims.ny_e > 0 %}
+    std::array<double, {{ model.name | upper }}_NYN> yrefN{};
+    {%- endif %}
+    {%- if dims.np > 0 %}
+    std::array<double, {{ model.name | upper }}_NP> p{};
+    {%- endif %}
+
+    {
+        std::scoped_lock lock(data_mutex_);
+        x0 = current_x_;
+        {%- if dims.ny_0 > 0 %}
+        yref0 = current_yref_0_;
+        {%- endif %}
+        {%- if dims.ny > 0 %}
+        yref = current_yref_;
+        {%- endif %}
+        {%- if dims.ny_e > 0 %}
+        yrefN = current_yref_e_;
+        {%- endif %}
+        {%- if dims.np > 0 %}
+        p = current_p_;
+        {%- endif %}
+    }
+
+    // Update solver
+    this->set_x0(x0.data());
+
+    {%- if dims.ny_0 > 0 %}
+    this->set_yref0(yref0.data());
+    {%- endif %}
+    {%- if dims.ny > 0 %}
+    this->set_yrefs(yref.data());
+    {%- endif %}
+    {%- if dims.ny_e > 0 %}
+    this->set_yref_e(yrefN.data());
+    {%- endif %}
+    {%- if dims.np > 0 %}
+    this->set_ocp_parameters(p.data(), p.size());
+    {%- endif %}
+
+    // Solve OCP
+    {%- if solver_options.nlp_solver_type == "SQP_RTI" %}
+    int status;
+    if (first_solve_) {
+        this->warmstart_solver_states(x0.data());
+        status = this->ocp_solve();
+    }
+    else {
+        status = this->feedback_rti_solve();
+    }
+
+    {%- else %}
+    int status = this->ocp_solve();
+    {%- endif %}
+    this->solver_status_behaviour(status);
+}
+
+void {{ ClassName }}::solver_status_behaviour(int status) {
+    if (status == ACADOS_SUCCESS) {
+        this->get_input(u0_.data(), 0);
+        this->publish_input(u0_, status);
+        {%- if solver_options.nlp_solver_type == "SQP_RTI" %}
+        first_solve_ = false;
+        this->prepare_rti_solve();
+        {%- endif %}
+    } else {
+        this->publish_input(u0_default_, status);
```

Comment:
I did not see what `u0_default_` is set to. I think it would be better to always return the `u0` from the solver. E.g. if the solver has returned max. number of iterations, the solution can still be useful.
Maybe one would additionally want to return the residuals to check how good the solution is.